crates/generator/src/typed_transaction/types.rs

```diff
@@ -0,0 +1,144 @@
+use std::convert::TryInto;
+
+use ckb_vm::Bytes;
+use gw_common::builtins::CKB_SUDT_ACCOUNT_ID;
+use gw_types::{
+    core::AllowedContractType,
+    packed::{ETHAddrRegArgs, L2Transaction, MetaContractArgs, SUDTArgs},
+    prelude::*,
+};
+
+use crate::error::{AccountError, TransactionValidateError};
+
+/// Types Transaction
+pub enum TypedTransaction {
+    EthAddrReg(EthAddrRegTx),
+    Meta(MetaTx),
+    SimpleUDT(SimpleUDTTx),
+    Polyjuice(PolyjuiceTx),
+}
+
+impl TypedTransaction {
+    pub fn from_tx(
+        tx: L2Transaction,
+        type_: AllowedContractType,
+    ) -> Result<Self, TransactionValidateError> {
+        let tx = match type_ {
+            AllowedContractType::EthAddrReg => Self::EthAddrReg(EthAddrRegTx(tx)),
+            AllowedContractType::Meta => Self::Meta(MetaTx(tx)),
+            AllowedContractType::Sudt => Self::SimpleUDT(SimpleUDTTx(tx)),
+            AllowedContractType::Polyjuice => Self::Polyjuice(PolyjuiceTx(tx)),
+            AllowedContractType::Unknown => return Err(AccountError::UnknownScript.into()),
+        };
+        Ok(tx)
+    }
+
+    /// Got expect cost of the tx, (transfer value + fee).
+    /// returns none if tx has no cost, it may happend when we call readonly interface of some Godwoken builtin contracts.
+    pub fn cost(&self) -> Option<u64> {
+        match self {
+            Self::EthAddrReg(tx) => tx.cost(),
+            Self::Meta(tx) => tx.cost(),
+            Self::SimpleUDT(tx) => tx.cost(),
+            Self::Polyjuice(tx) => tx.cost(),
+        }
+    }
+}
+
+pub struct EthAddrRegTx(L2Transaction);
+
+impl EthAddrRegTx {
+    pub fn cost(&self) -> Option<u64> {
+        use gw_types::packed::ETHAddrRegArgsUnion::*;
+
+        let args: Bytes = self.0.raw().args().unpack();
+        let args = ETHAddrRegArgs::from_slice(&args).ok()?;
```

Comment:
molecule structures is implemented as 'zero-copy' under the 'std' rust environment, only the reference of a pointer is cloned.

I choose to use non reader version to avoid too many lifetime notations.

Reply:
My reading is that `Reader::from_slice` and `Entity::new_unchecked` are zero-copy, but `Entity::from_slice` or `Reader::to_entity` are not:

```rust
impl<'r> molecule::prelude::Reader<'r> for SUDTArgsReader<'r> {
    fn to_entity(&self) -> Self::Entity {
        Self::Entity::new_unchecked(self.as_slice().to_owned().into())
    }
}
```

Lifetime annotations are not necessary. See https://github.com/nervosnetwork/godwoken/pull/626/files#diff-737eb91a5e5aa0d1609b8bd48181fd1b9d647de85f773ad585172da442e654d7